### PR TITLE
Fix the categories tree

### PR DESCRIPTION
**Pull request by hibatallahAouadni**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | 1.6.1.x
| Description?  | the problem is that category will be deleted when you try to add main cat to sub-category.
| Type?         | bug fix
| Category?     | BO
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | http://forge.prestashop.com/browse/PSCSX-9097
| How to test?  | try to edit a category, you will find all its sub-categories are disabled, so you can't add to one of its sub-cat.

### CODE REVIEW

**Comment by marionf** [comment]
This PR is solving a problem with the category tree on product page due to the PR #7925 
Thanks @hibatallahAouadni for solving it.

**Comment by maximebiloe** [comment]
On the original PR (https://github.com/PrestaShop/PrestaShop/pull/7925), you made some changes in PHP. Is it wanted to not doing them here?

File: classes/helper/HelperTreeCategories.php [hunk 1]
```diff
@@ -59,17 +59,45 @@ public function __construct($id, $title = null, $root_category = null,
...
+                    $categories[$childId] = array($child);
+                }
+
+                foreach ($tree[$categoryId]['children'] as $childId => $child) {
+                    $subtree = $this->fillTree($categories, $childId);
+
+                    foreach ($subtree as $subcategoryId => $subcategory) {
+                        $tree[$categoryId]['children'][$subcategoryId] = $subcategory;
+                    }
+                }
```
Comment (hibatallahAouadni) [inline classes/helper/HelperTreeCategories.php]: @maximebiloe from 65 to 97 all these are the changement made by thierry (in the branch 1.7.0.x and not 1.6.1.x)

**Comment by maximebiloe** [comment]
Thank you @hibatallahAouadni

**Comment by Artxlv** [comment]
This fix causes very slow loading Associations Tab, when you want to edit product. I upgraded from 1.6.1.11 to 1.6.1.17 and could not get a category tree on product editing page anymore - I always got the timeout error. So I traced until the HelperTreeCategories.php fillTree method and found this bad fix. With old fillTree method categories loaded more than 10 times faster. (I have ~35000 categories)

**Comment by hibatallahAouadni** [comment]
the PHP part is Thierry PR: [#6318](https://github.com/PrestaShop/PrestaShop/pull/6318) and I just "cherry-picked"

**Comment by Artxlv** [comment]
Sorry, forwarded.

**Pull request merged**